### PR TITLE
[Security] Be able to know the reasons of the denied access 

### LINKED ISSUE #27995: [Security][DX] Be able to know why exactly SecurityVoter returns false
Sometimes I find it hard to debug why `is_granted` returns `true`/`false` in some cases. I know there is at least one feature request (#26343) that tries to improve this, but #24501 inspired me to create this feature request.

[This](https://symfony.com/doc/current/security/voters.html) code:

    // src/Security/PostVoter.php
    if (! $user->hasRole('ROLE_ADMIN')) {
        return false;
    }

    // src/Controller/PostController.php
    $this->denyAccessUnlessGranted('view', $post);

Reminds me of this:
Computer says no

It would be great to re-use the idea behind `TransitionBlocker` when using [Voters](https://symfony.com/doc/current/security/voters.html). Instead, one should be able to write something like

    if (! $user->hasRole('ROLE_ADMIN')) {
        yield new VoterBlocker('You need admin rights to do this');
    }

This has two benefits:
* This will make it easier to debug because the reason can be shown in the Profiler. This will extend the [Detailed security voters decisions in Symfony 4.2](https://symfony.com/blog/new-in-symfony-4-2-detailed-security-voters-decisions)
* There could be a way to use this message to inform the user why it can't perform an action. Using `is_granted` feels like implementing "Computer says no" instead of "Sorry, you need to be an admin/this post is private/etc/". This can work similar to [Transition Blockers](https://symfony.com/blog/new-in-symfony-4-1-workflow-improvements#added-transition-blockers) in Symfony's Workflow Component 4.1 and the recently added `workflow_transition_blockers` from [Symfony 4.3](https://symfony.com/blog/new-in-symfony-4-3-workflow-improvements#added-workflow-transition-blockers-twig-function).

### PR DESCRIPTION
**Pull request by azjezz**
| Q             | A
| ------------- | ---
| Branch?       | 5.x 
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | yes
| Tickets       | Fix #27995, #26343, #35592
| License       | MIT
| Doc PR        | n/a

This is a PR continuing the work started by @maidmaid and @noniagriconomie in #35592

---

About the voters, the gif of the the original issue sums up the situation well :

no

Currently, the voters can only say yes or no. This PR allows to give a reason from the voters.

Before :

```php
class PostVoter extends Voter
{
    // ...

    protected function voteOnAttribute($attribute, $subject, TokenInterface $token)
    {
        if ($subject->getAuthor() !== $token->getUser()) {
            return false;
        }

        if ($subject->getStatus() !== 'draft') {
            return false;
        }

        return true;
    }
}
```

After:
```php
class PostVoter extends Voter
{
    // ...

    protected function voteOnAttribute($attribute, $subject, TokenInterface $token)
    {
        if ($subject->getAuthor() !== $token->getUser()) {
            return $this->deny('You are not the author.')
        }

        if ($subject->getStatus() !== 'draft') {
            return $this->deny('The post is no more a draft.')
        }

        return $this->grant();
    }
}
```

A voter returns no more an int based on the constant ACCESS_{GRANTED,ABSTAIN,DENIED} but a Vote object which contains the access result and the reason. Thus, the security panel of the profiler can display the reason of each vote.

security

The access decision manager returns no more a bool for the final verdict but an AccessDecision object which contains the final verdict and all the votes. Thus, the 403 response can display all the denied raisons.

security-ex

---

TODO:
  - [x] add BC Layer
  - [x] migrate existing voters
  - [ ] update old tests
  - [ ] add new tests

### CODE REVIEW

**Comment by noniagriconomie** [comment]
Tank you for continuing this @azjezz I'll review very soon
Also to be mentioned, all the initial work has been done by @maidmaid, I've just done some reviews like others ppl :)

File: src/Symfony/Component/Security/Core/Authorization/Voter/VoterInterface.php [hunk 1]
```diff
@@ -33,7 +33,7 @@ interface VoterInterface
      * @param mixed $subject    The subject to secure
      * @param array $attributes An array of attributes associated with the method being invoked
      *
-     * @return int either ACCESS_GRANTED, ACCESS_ABSTAIN, or ACCESS_DENIED
+     * @return int|Vote either ACCESS_GRANTED, ACCESS_ABSTAIN, or ACCESS_DENIED
```
Comment (noniagriconomie) [inline src/Symfony/Component/Security/Core/Authorization/Voter/VoterInterface.php:36]: related to comment on `Voter/Voter.php`

File: src/Symfony/Component/Security/Core/Authorization/Voter/Vote.php [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+ * It can also contains a reason explaining the vote decision.
+ *
+ * @author Dany Maillard <danymaillard93b@gmail.com>
+ */
+final class Vote
+{
+    use AccessTrait;
+
+    private $reason;
+    private $parameters;
```
Comment (noniagriconomie) [inline src/Symfony/Component/Security/Core/Authorization/Voter/Vote.php:25]: i would call this `$context` like in workflow, wdyt?

File: src/Symfony/Component/Security/Core/Authorization/AccessDecisionManager.php [hunk 3]
```diff
@@ -77,28 +83,29 @@ public function decide(TokenInterface $token, array $attributes, $object = null/
...
+            if ($vote->isGranted()) {
+                return AccessDecision::createGranted($votes);
             }
 
-            if (VoterInterface::ACCESS_DENIED === $result) {
+            if ($vote->isDenied()) {
                 ++$deny;
-            } elseif (VoterInterface::ACCESS_ABSTAIN !== $result) {
+            } elseif (!$vote->isAbstain()) {
                 trigger_deprecation('symfony/security-core', '5.3', 'Returning "%s" in "%s::vote()" is deprecated, return one of "%s" constants: "ACCESS_GRANTED", "ACCESS_DENIED" or "ACCESS_ABSTAIN".', var_export($result, true), get_debug_type($voter), VoterInterface::class);
```
Comment (OskarStark) [inline src/Symfony/Component/Security/Core/Authorization/AccessDecisionManager.php]:
```suggestion
                trigger_deprecation('symfony/security-core', '5.3', 'Returning "%s" in "%s::vote()" is deprecated, return one of "%s" constants: "ACCESS_GRANTED", "ACCESS_DENIED" or "ACCESS_ABSTAIN".', var_export($votes, true), get_debug_type($voter), VoterInterface::class);
```

?

Comment (azjezz) [reply src/Symfony/Component/Security/Core/Authorization/AccessDecisionManager.php]: this condition needs to be moved down to vote method, still WIP 😄

File: src/Symfony/Component/Security/Core/Authorization/AccessDecisionManager.php [hunk 4]
```diff
@@ -206,4 +216,22 @@ private function decidePriority(TokenInterface $token, array $attributes, $objec
 
         return $this->allowIfAllAbstainDecisions;
     }
+
+    private function decideIfAllAbstainDecisions(): AccessDecision
```
Comment (noniagriconomie) [inline src/Symfony/Component/Security/Core/Authorization/AccessDecisionManager.php]: i would directly write this code, the function seem not mandatory

File: src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php [hunk 5]
```diff
@@ -75,7 +108,7 @@ abstract protected function supports(string $attribute, $subject);
      *
      * @param mixed $subject
      *
-     * @return bool
+     * @return Vote Returning a boolean is deprecated since Symfony 5.1. Return a Vote object instead.
```
Comment (noniagriconomie) [inline src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php:111]: typo, 5.3
also, shouldnt we create someting like https://github.com/symfony/symfony/pull/35592#issuecomment-693440193 to make it BC?

**Review by noniagriconomie — COMMENTED** [review]
some early reviews

**Comment by noniagriconomie** [comment]
👋 ping @azjezz will you be able to rebase and finish for v5.4? thx

**Comment by fabpot** [comment]
Let's close for now as there is no more work here. @azjezz Feel free to reopen if you'd like to finish it.

**Comment by yellow1912** [comment]
I would really like to see this feature merged. Is it possible for me to pick up the work? @fabpot @azjezz I can use the current 5.4 base, merge the changes made.

**Comment by azjezz** [comment]
@yellow1912 feel free to do so, I'm currently unable to do so due to lack of time :)

**Pull request closed without merging**